packages/task/src/browser/task-terminal-widget-manager.ts

```diff
@@ -0,0 +1,166 @@
+
+import { inject, injectable, postConstruct } from 'inversify';
+import { ApplicationShell, WidgetOpenerOptions } from '@theia/core/lib/browser';
+import { TerminalWidget } from '@theia/terminal/lib/browser/base/terminal-widget';
+import { TerminalWidgetFactoryOptions } from '@theia/terminal/lib/browser/terminal-widget-impl';
+import { TerminalService } from '@theia/terminal/lib/browser/base/terminal-service';
+import { TaskConfiguration, TaskWatcher, TaskExitedEvent, PanelKind } from '../common';
+import { TaskDefinitionRegistry } from './task-definition-registry';
+
+export interface TaskTerminalWidgetOpenerOptions extends WidgetOpenerOptions, TerminalWidgetFactoryOptions {
+    taskId: number;
+    taskPanel?: PanelKind;
+    taskConfig?: TaskConfiguration;
+}
+export namespace TaskTerminalWidgetOpenerOptions {
+    export function isDedicatedTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return !!options.taskPanel && options.taskPanel === PanelKind.Dedicated;
+    }
+
+    export function isNewTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return !!options.taskPanel && options.taskPanel === PanelKind.New;
+    }
+
+    export function isSharedTerminal(options: TaskTerminalWidgetOpenerOptions): boolean {
+        return options.taskPanel === undefined || options.taskPanel === PanelKind.Shared;
+    }
+}
+
+@injectable()
+export class TaskTerminalWidgetManager {
+
+    @inject(ApplicationShell)
+    protected readonly shell: ApplicationShell;
+
+    @inject(TaskDefinitionRegistry)
+    protected readonly taskDefinitionRegistry: TaskDefinitionRegistry;
+
+    @inject(TerminalService)
+    protected readonly terminalService: TerminalService;
+
+    @inject(TaskWatcher)
+    protected readonly taskWatcher: TaskWatcher;
+
+    // Map indexed by terminal widget id
+    protected terminalWidgetMap: Map<string, { isDedicated: boolean, taskId: number, taskConfig?: TaskConfiguration, isIdle: boolean }> = new Map();
+
+    @postConstruct()
+    protected init(): void {
+        this.taskWatcher.onTaskExit((event: TaskExitedEvent) => {
+            const finishedTaskId = event.taskId;
+            // find the terminal where the task ran, and mark it as "idle"
+            for (const widgetId of this.terminalWidgetMap.keys()) {
+                const entry = this.terminalWidgetMap.get(widgetId)!;
+                if (entry.taskId === finishedTaskId) {
+                    entry.isIdle = true;
+
+                    const terminalWidget = this.terminalService.getById(widgetId);
+                    if (terminalWidget) {
+                        terminalWidget.scrollToBottom();
+                        terminalWidget.writeLine('\x1b[1m\n\rTerminal will be reused by tasks. \x1b[0m\n');
+                    }
+
+                    break;
+                }
+            }
+        });
+    }
+
+    async getOrCreateWidget(options: TaskTerminalWidgetOpenerOptions): Promise<TerminalWidget> {
```

Comment:
I would call it `open` since it does not just create a terminal, but actually opens it, also ti accepts opener options.